### PR TITLE
Add MariaDB to AWS RDS auto discovery

### PR DESCRIPTION
**Pull request by jakule**
This PR adds auto-discovery to MariaDB RDS including docs changes. The minimum supported version is MariaDB 10.6 as AWS doesn't support IAM on older versions: https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html

### CODE REVIEW

File: lib/services/database.go [hunk 1]
```diff
@@ -452,9 +476,9 @@ const (
 	// RDSEndpointTypeReader is the endpoint that load-balances connections across the Aurora Replicas that are
 	// available in a RDS cluster.
 	RDSEndpointTypeReader RDSEndpointType = "reader"
-	// RDSEndpointTypeCustom is the endpoint that specifieds one of the custom endpoints associated with the RDS cluster.
+	// RDSEndpointTypeCustom is the endpoint that specified one of the custom endpoints associated with the RDS cluster.
```
Comment (codingllama) [inline lib/services/database.go:479]: Maybe? The sentence seems off.

```suggestion
	// RDSEndpointTypeCustom is the endpoint that specifies one of the custom endpoints associated with the RDS cluster.
```

File: lib/services/database_test.go [hunk 2]
```diff
@@ -364,6 +364,54 @@ func TestIsRDSClusterSupported(t *testing.T) {
...
+		t.Run(test.name, func(t *testing.T) {
+			cluster := &rds.DBInstance{
+				DBInstanceArn:       aws.String("arn:aws:rds:us-east-1:1234567890:instance:test"),
+				DBClusterIdentifier: aws.String(test.name),
+				DbiResourceId:       aws.String(uuid.New().String()),
+				Engine:              aws.String(test.engine),
+				EngineVersion:       aws.String(test.engineVersion),
+			}
+
+			require.Equal(t, test.isSupported, IsRDSInstanceSupported(cluster))
```
Comment (codingllama) [inline lib/services/database_test.go:410]: nit: Add a message to the error, the default Equal message is not great for booleans.

```suggestion
			got, want := IsRDSInstanceSupported(cluster), test.isSupported
			require.Equal(t, want, got, "IsRDSInstanceSupported = %v, want = %v", got, want)
```

**Review by codingllama — APPROVED** [review]

File: docs/pages/database-access/guides/rds.mdx [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ---
-title: Database Access with AWS RDS and Aurora for PostgreSQL and MySQL
-description: How to configure Teleport Database Access with AWS RDS and Aurora for PostgreSQL and MySQL.
+title: Database Access with AWS RDS and Aurora for PostgreSQL, MySQL and MariaDB.
```
Comment (ptgott) [inline docs/pages/database-access/guides/rds.mdx:2]:
```suggestion
title: Database Access with AWS RDS and Aurora for PostgreSQL, MySQL and MariaDB
```

Comment (ptgott) [reply docs/pages/database-access/guides/rds.mdx:2]: Also, the `title` is now long enough to be truncated in Google's search results. To prevent this, let's add an `h1` frontmatter item with this title—this will just be used for presentation within the page—and give `title` the value, `Database Access with AWS RDS and Aurora`.

File: docs/pages/database-access/guides/rds.mdx [hunk 4]
```diff
@@ -14,6 +14,11 @@ This guide will help you to:
   writing so it can't be used with Database Access.
 </Admonition>
 
+<Admonition type="note" title="MariaDB supported versions">
```
Comment (ptgott) [inline docs/pages/database-access/guides/rds.mdx:17]: The two back-to-back Admonitions are kind of hard to follow. I think we should include a single Admonition that starts with a general statement like, "The following products are not compatible with Database Access," followed by a two-item list that includes the products mentioned in these Admonitions.

File: docs/pages/database-access/guides/rds.mdx [hunk 5]
```diff
@@ -318,8 +323,8 @@ Access for RDS. See below how to enable it for your database engine.
   GRANT rds_iam TO alice;
   ```
   </TabItem>
-  <TabItem label="MySQL">
-  MySQL users must have RDS authentication plugin enabled:
+  <TabItem label="MySQL/MariaDB">
+  MySQL and MariaDB users must have RDS authentication plugin enabled:
```
Comment (ptgott) [inline docs/pages/database-access/guides/rds.mdx:327]:
```suggestion
  MySQL and MariaDB users must have the RDS authentication plugin enabled:
```

File: docs/pages/database-access/guides/rds.mdx [hunk 6]
```diff
@@ -14,6 +14,11 @@ This guide will help you to:
   writing so it can't be used with Database Access.
 </Admonition>
 
+<Admonition type="note" title="MariaDB supported versions">
+  The minimum supported version of MariaDB is 10.6. Older versions don't support
```
Comment (r0mant) [inline docs/pages/database-access/guides/rds.mdx:18]: Nit: I would clarify that this is specifically for RDS MariaDB. Might be clear enough as it is given that this is RDS specific guide but I wonder if somebody might get confused by it anyway.

File: lib/services/database.go [hunk 7]
```diff
@@ -369,6 +370,27 @@ func rdsTagsToLabels(tags []*rds.Tag) map[string]string {
...
+	ver, err := semver.NewVersion(aws.StringValue(instance.EngineVersion))
+	if err != nil {
+		log.Errorf("Failed to parse RDS MariaDB version: %s", aws.StringValue(instance.EngineVersion))
+		return false
+	}
+
+	// Min supported MariaDB version that supports IAM is 10.6
+	// https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html
+	minIAMSupportedVer := semver.New("10.6.0")
+	return ver.Compare(*minIAMSupportedVer) >= 0
```
Comment (r0mant) [inline lib/services/database.go:391]: Nit: I think using `LessThan` is a bit more readable.

File: lib/srv/db/cloud/watchers/rds.go [hunk 8]
```diff
@@ -106,6 +106,14 @@ func (f *rdsDBInstancesFetcher) getRDSDatabases(ctx context.Context) (types.Data
 	}
 	databases := make(types.Databases, 0, len(instances))
 	for _, instance := range instances {
+		if !services.IsRDSInstanceSupported(instance) {
```
Comment (r0mant) [inline lib/srv/db/cloud/watchers/rds.go:109]: What about Aurora clusters? Does Aurora provide MariaDB engine?

Comment (jakule) [reply lib/srv/db/cloud/watchers/rds.go:109]: I don't see MariaDB in Aurora (checked website + docs).

File: lib/services/database.go [hunk 9]
```diff
@@ -369,6 +370,27 @@ func rdsTagsToLabels(tags []*rds.Tag) map[string]string {
 	return labels
 }
 
+// IsRDSInstanceSupported returns true if database supports IAM authentication.
+// Currently, only MariaDB is being checked as all other RDS databases supports
+// IAM authentication in all configurations.
+func IsRDSInstanceSupported(instance *rds.DBInstance) bool {
```
[the remark this answers is not in the document]
Comment (jakule) [reply lib/services/database.go:376]: @greedy52 You're right. I wanted to say, that the check haven't been implemented for other databases and MariaDB is currently the only one that is being checked. I'll change the comment. In terms of adding support for other databases, I think we could do that, but if we do we should do it in as a separate PR.
Also there is a separate list for Postgres.

**Comment by jakule** [comment]
Checking the updated docs page I noticed something. Here is a screenshot
Screen Shot 2022-03-04 at 6 07 58 PM
So the page is targeting Teleport 9 release and everything described here (including support for RDS MariaDB) applies to v9, except `Install Teleport 7.0.0`. In my opinion is confusing. Any thoughts? @r0mant @ptgott

**Review by greedy52 — APPROVED** [review]

File: lib/services/database.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (greedy52) [reply lib/services/database.go]: Thanks!

**Comment by greedy52** [comment]
https://github.com/gravitational/teleport/blob/b8cae1abfefd708e3c9f34a991481551972a6b91/docs/config.json#L814-L823

We haven't updated the version for docs?

File: docs/pages/database-access/guides/rds.mdx [hunk 10]
```diff
@@ -1,6 +1,6 @@
 ---
-title: Database Access with AWS RDS and Aurora for PostgreSQL and MySQL
-description: How to configure Teleport Database Access with AWS RDS and Aurora for PostgreSQL and MySQL.
+title: Database Access with AWS RDS and Aurora for PostgreSQL, MySQL and MariaDB
```
Comment (ptgott) [inline docs/pages/database-access/guides/rds.mdx:2]: To prevent this title from being truncated in Google's search results, we should get this to around 55 chars or less, e.g., by making it "Database Access with AWS RDS and Aurora." We can then add an `h1` attribute to the frontmatter that uses the original title, "Database Access with AWS RDS and Aurora for PostgreSQL, MySQL and MariaDB." The `h1` attribute is not used to populate the page's `title` element, while `title` is.

File: docs/pages/database-access/guides/rds.mdx [hunk 11]
```diff
@@ -9,9 +9,10 @@ This guide will help you to:
 - Set up Teleport to access your RDS instances and Aurora clusters.
 - Connect to your databases through Teleport.
 
-<Admonition type="note" title="Aurora Serverless">
-  Aurora Serverless does not support IAM authentication at the time of this
-  writing so it can't be used with Database Access.
+<Admonition type="note" title="Supported versions">
+    The following products are not compatible with Database Access as they don't support IAM authentication:
+    * Aurora Serverless.
+    * RDS MariaDB in version lower than 10.6.
```
Comment (ptgott) [inline docs/pages/database-access/guides/rds.mdx:15]:
```suggestion
    * RDS MariaDB in versions lower than 10.6.
```

Comment (jakule) [reply docs/pages/database-access/guides/rds.mdx]: Changed, thx.

File: docs/pages/database-access/guides/rds.mdx (hunk 10 above)
Comment (jakule) [reply docs/pages/database-access/guides/rds.mdx]: Good tip. I'd never think about it. Changed.

**Review by ptgott — APPROVED** [review]

File: docs/pages/database-access/guides/rds.mdx [hunk 12]
```diff
@@ -9,9 +10,10 @@ This guide will help you to:
 - Set up Teleport to access your RDS instances and Aurora clusters.
 - Connect to your databases through Teleport.
 
-<Admonition type="note" title="Aurora Serverless">
-  Aurora Serverless does not support IAM authentication at the time of this
-  writing so it can't be used with Database Access.
+<Admonition type="note" title="Supported versions">
+    The following products are not compatible with Database Access as they don't support IAM authentication:
+    - Aurora Serverless.
+    - RDS MariaDB in versions lower than 10.6.
```
Comment (r0mant) [inline docs/pages/database-access/guides/rds.mdx:16]: Nit: "in versions" reads strange
```suggestion
    - RDS MariaDB versions lower than 10.6.
```

**Review by r0mant — APPROVED** [review]
